### ISSUE TITLE
Checking for methods defined in shipped extensions

### ISSUE THREAD

**Issue by Eneroth3** [SketchUp/rubocop-sketchup#31]
It could be helpful to developers to get a warning/error if their extensions rely on methods defined in shipped extensions. I have myself accidentally used some of these methods without realizing they weren't a part of the API and later got errors when running the extension on a computer where these extensions were disabled.

**Comment by Eneroth3** [comment]
Here are the methods extending base classes defined by Dynamic Components and Advanced Camera Tools. Are there other shipped extensions I should check?

# Dynamic Components

## Sketchup::Model

* typename
* entityID
* delete_attribute
* layer

## Sketchup::Drawingelement

* local_transformation
* scaled_size
* unscaled_size
* set_last_size
* last_scaling_factors

## Geom::Transformation

* rotx
* roty
* rotz
* xscale
* yscale
* zscale

## UI::WebDialog

* last_width=
* last_height=
* last_width
* last_height

# Advanced Camera Tools

## Sketchup::Entity

* get_attributes
* has_attributes?

**Comment by thomthom** [comment]
That's a good idea. I can scan the source of all our shipped and check if there's anything missing.

**Comment by thomthom** [comment]
Some of this might not be trivial to catch. It's not always trivial to figure out the type of an object with static analysis in Ruby.

For the methods on Model which dupes Entity it would be near by impossible to detect unless we inferred some hints by the variable name (model.entityID). But that might not catch where it matters - as the use of these methods are often in generic utility methods.

For the other methods were their names are unique we could catch them when they are called (on anything.) That could lead to some false positives, but the purpose of this cop would be to make developers aware of potential issue. It can be disabled if needed by the user.

**Comment by DanRathbun** [comment]
All of these methods were good (monkey patch) additions to the API, and we've asked forever (especially the **`Geom::Transformation`** property getters) that these be "baked" into the API.

It's long past time. (The DC extension can be edited to first do duck typing on the above classes before it does the monkey patching. Or, even better, start using refinements for newer SU versions. They are no longer experimental and have gotten a lot of updates for Ruby 2.4.0, including the new ability to refine modules as well as classes.)

Anyway, adding them to the API, would eliminate **this** issue going forward. (Slowly disappears as a problem as more and more users move to newer SU editions running newer Ruby versions.)

Another option, is to fix the DC code so (for example) it does not need the "dummy" methods it added to the **`Sketchup::Model`** class just to avoid "duck typing" within many of it's loop constructs.

**Comment by DanRathbun** [comment]
Don't get me wrong, this "idea" of Julia's for rubocop is a good one.

> @Eneroth3  rotx, toty, and rotz are just awful methods that returns strings representing angles in degrees rounded to integers. Proper methods for getting the angles in the radians with adequate precision would be quite useful but I would not want these methods in their current state to be added to the API. Formatting strings is something you can do yourself when printing values to the user, not something the API should do when you try to access data from it.
> 
> xscale, yscale and zscale would be useful additions though, especially if accompanied with corresponding setter methods. Why not setters for individual axes too?

Sounds like you should be the one to write up API FR(s) for the **`Geom::Transformation`** class ...
https://github.com/SketchUp/api-issue-tracker/issues
... and add your determinant method, and those nifty non-normalized axis getters (as bang methods? ie, **`xaxis!()`**, **`yaxis!()`** and **`zaxis!()`**.)

> However I don't think the problem would be solved by forcing these methods, as they were originally designed for these plugins, into the API. **At least not all of them.**

**Oh, for sure not all of them.** Like I said with the model object getters, basically Scott was lazy and added those so a model would act like a component instance in his loop constructs (ie, the model object would not throw **`NoMethodError`**.)

> Some are extremely inconsistent to the API, some just weird (bad weird).

Hehe, like **`Model#layer()** perhaps ? **Take the BEST, forget the REST.**

> A better solution IMO is to recommend extension developers to disable these "dirty" extensions.

Coders would be better off using a mixin module of these methods (or better written like your libs) that they can mix into their own classes and extension modules. **Or perhaps refinement.**

> Or maybe update these extension to check the caller for each of these method and raise an error/warn if its from an external extension.

**Doable**, but expensive speedwise. It would slow DC way down. (But if DC was fixed to not use these monkey patches, then there'd be no down side for DC at least.)

**Comment by Eneroth3** [comment]
Even if only the best of these methods would be added to the API there could still be problems with developers accidentally relying on the others, thinking they are in the API, so it wouldn't solve the problem.

Some could be added anyway, but the question is where to draw the line for what belongs in the API. At true axes? At the determinant? At Euler angles? At extracting shearing as a separate transformation? At getting the scale factor for a specific plane? In a way the API is already fulfilling it's duty of giving access to the data when it gives you the raw transformation matrix. I can understand if the extensibility team don't want to create and maintain code in their codebase for working with the transformation. That's where a separate library comes neatly into the picture!

Anyhow, this is another discussion.

For the original issue I'm thinking about making a really small extension that monkey patches the monkey patched extensions, and prints a warning to the console whenever these methods are used. Merely publishing and be able to refer to such a plugin could help the developer community to realize this is a potential issue that should be taken into account when develping.

**Comment by DanRathbun** [comment]
> I can understand if the extensibility team don't want to create and maintain code in their codebase for working with the transformation. 

They have enough problems getting the current feature set properly documented it seems.

> That's where a separate library comes neatly into the picture!

Then either it should be a refined class or a real subclass of **`Geom::Transformation`** (loaded** from ONE shared community project, This is necessary especially for refinements where method overrides are  allowed only once. I've had a skeleton of this project for several years.) 

> Anyhow, this is another discussion.

Yup. we've gone off the "road" a bit here. Back on the trail ...

> For the original issue I'm thinking about making a really small extension that monkey patches the monkey patched extensions, and prints a warning to the console whenever these methods are used.

I "half-like" the idea as an extension. Ii's a good idea, but I'd prefer it be part of a developers tool like rubocop or TestUp, and that it be optional. It could be switched off, but switched on as a default.

**In other words "monkey patching" to produce a warning on every call, continuously whilst a user is trying to do work, is a no-no. It has more merit as part of a test suite.**

Another "item" is that you'd need to do introspection to determine if the API has "baked in" each method before spitting out a warning.
```ruby
Sketchup::Model.instance_method(:typename).source_location
# returns nil for native C method definitions:
```
Not a total guarantee as coders can also do C extensions, but ...

**Comment by DanRathbun** [comment]
@Eneroth3 @thomthom My quick a dirty MonkeyPatchChecker found a few more ...
[MonkeyPatchChecker.zip](https://github.com/SketchUp/rubocop-sketchup/files/1583328/MonkeyPatchChecker.zip)

```ruby
MPC.list_monkey_patches

 -------Module/Class--------  --type--  --method------------  line  --filepath------------------------------------------
 Geom::Transformation         instance  rotx                   117  /su_dynamiccomponents/ruby/dcutils.rbe              
 Geom::Transformation         instance  roty                   136  /su_dynamiccomponents/ruby/dcutils.rbe              
 Geom::Transformation         instance  rotz                   159  /su_dynamiccomponents/ruby/dcutils.rbe              
 Geom::Transformation         instance  xscale                 183  /su_dynamiccomponents/ruby/dcutils.rbe              
 Geom::Transformation         instance  yscale                 196  /su_dynamiccomponents/ruby/dcutils.rbe              
 Geom::Transformation         instance  zscale                 209  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Camera             instance  copy                  2268  /su_advancedcameratools/advancedcameratools_main.rbs
 Sketchup::Camera             instance  update                2287  /su_advancedcameratools/advancedcameratools_main.rbs
 Sketchup::Camera             instance  same_camera_params?   2307  /su_advancedcameratools/advancedcameratools_main.rbs
 Sketchup::ComponentInstance  instance  get_attributes         145  /su_advancedcameratools/advancedcameratools_main.rbs
 Sketchup::ComponentInstance  instance  has_attributes?        158  /su_advancedcameratools/advancedcameratools_main.rbs
 Sketchup::ComponentInstance  instance  copy                   289  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::ComponentInstance  instance  description            300  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Drawingelement     instance  local_transformation   343  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Drawingelement     instance  scaled_size            401  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Drawingelement     instance  unscaled_size          423  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Drawingelement     instance  set_last_size          451  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Drawingelement     instance  last_scaling_factors   468  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Entity             instance  get_attributes         126  /su_advancedcameratools/advancedcameratools_main.rbs
 Sketchup::Entity             instance  has_attributes?        134  /su_advancedcameratools/advancedcameratools_main.rbs
 Sketchup::Model              instance  typename               496  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Model              instance  entityID               507  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Model              instance  delete_attribute       519  /su_dynamiccomponents/ruby/dcutils.rbe              
 Sketchup::Model              instance  layer                  534  /su_dynamiccomponents/ruby/dcutils.rbe              
 UI::WebDialog                instance  last_width=            245  /su_dynamiccomponents/ruby/dcutils.rbe              
 UI::WebDialog                instance  last_height=           250  /su_dynamiccomponents/ruby/dcutils.rbe              
 UI::WebDialog                instance  last_width             255  /su_dynamiccomponents/ruby/dcutils.rbe              
 UI::WebDialog                instance  last_height            260  /su_dynamiccomponents/ruby/dcutils.rbe              
```

**Comment by thomthom** [comment]
I just added some compatibility checks for the API in #4. Checking for usage of modules, classes and constants are easy as you have enough information to determine the full type name.

With method calls it's more difficult, as it's not possible to reliably determine the receiver of the method call. But with the `SketchupSuggestions/Compatibility` I added some basic method checks by testing for a subset of the method names which were unlikely to be ambiguous. Could try the same thing here with this list.

**Comment by thomthom** [comment]
Got a working prototype. Will need to test this for a while to confirm it's not too noisy.

![2018-03-07_16h24_56](https://user-images.githubusercontent.com/192418/37100789-2659f032-2224-11e8-91da-83e17de4ad67.png)
